Deprecate unsafe legacy format in Solovay Kitaev

Cryoris opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Deprecate the legacy `.npy` format for serializing the basic approximations in Solovay Kitaev. This internally relies on pickle and could be potentially unsafe.

Cryoris committed:
Deprecation warnings for legacy SK methods

Cryoris committed:
add reno, mention vulernability

Cryoris committed:
more warnings

jakelishman commented on qiskit/synthesis/discrete_basis/solovay_kitaev.py:
```diff
@@ -249,7 +260,6 @@ def query_basic_approximation(self, gate: np.ndarray | Gate) -> QuantumCircuit:
         since="2.1",
         additional_msg="Use query_basic_approximation instead, which takes a Gate or matrix "
         "as input and returns a QuantumCircuit object.",
-        pending=True,
     )
```
Need to change to "since 2.3" now, I think

jakelishman commented on test/python/transpiler/test_solovay_kitaev.py:
```diff
@@ -74,7 +74,8 @@ class TestSolovayKitaev(QiskitTestCase):
 
-        self.basic_approx = generate_basic_approximations([HGate(), TGate(), TdgGate()], 3)
+        with self.assertWarns(DeprecationWarning):
+            self.basic_approx = generate_basic_approximations([HGate(), TGate(), TdgGate()], 3)
```
Can you include the warning message in an `assertWarnsRegex`?  Even from this PR, I can't really see _why_ `generate_basic_approximations` should now raise a `DeprecationWarning` (isn't the only thing we newly raise related to _loading_ them?).

Cryoris replied:
I'll add the messages 👍🏻 I added the deprecation to both loading and generation (since nothing will be able to consume the deprecated format in the future)

jakelishman commented on test/python/transpiler/test_solovay_kitaev.py:
```diff
@@ -153,8 +154,9 @@ def test_str_basis_gates(self):
-        sk = SolovayKitaev(3, basic_approx)
+        with self.assertWarns(DeprecationWarning):
+            basic_approx = generate_basic_approximations(["h", "t", "s"], 3)
+            sk = SolovayKitaev(3, basic_approx)
```
Which of these two calls is suppose to raise?

Cryoris replied:
You called out my laziness -- both are 🙂 Fixed now!

jakelishman commented on test/python/transpiler/test_solovay_kitaev.py:
```diff
@@ -228,16 +234,18 @@ def test_load_from_file(self):
             # Run SK pass using gate_approx_library
-            reference = SolovayKitaev(basic_approximations=gate_approx_library)(circuit)
+            with self.assertWarns(DeprecationWarning):
+                reference = SolovayKitaev(basic_approximations=gate_approx_library)(circuit)
```
Which of the calls is supposed to raise?  (same question every time we have this pattern.)  Can we make the warning catch stricter?

Cryoris committed:
Fix deprecation version and more explicit tests

alexanderivrii commented on releasenotes/notes/deprecate-sk-legacy-format-7e2770aa302ce6d3.yaml:
```diff
@@ -0,0 +1,12 @@
+    Deprecated the legacy serialization format used in :class:`.SolovayKitaevDecomposition`,
+    which was based on pickling the basic approximations in form of a Python dictionary.
+    This loading process is a potential vulnerability and should only be used with trusted files
+    and the new format does not have this problem.
```
```suggestion
    The loading process is a potential security vulnerability and should only be used with trusted files.
    The new serialization format avoids this vulnerability.
```

alexanderivrii reviewed (commented):
Thanks for closing the door to this potential security vulnerability. I have two small comments:

* The function `generate_basic_approximations` is still used in the docstring examples for the `SolovayKitaev` transformation pass. Can we update the examples to avoid mentioning `generate_basic_approximations` but to show how one can store and load (binary) approximations between runs?

* I thought that we already raise an error if we try to save approximations in the "npy" format, it's only the loading part that's still allowed.

Cryoris committed:
Update releasenotes/notes/deprecate-sk-legacy-format-7e2770aa302ce6d3.yaml

Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

Cryoris committed:
Fix docs

Cryoris committed:
Merge branch 'sk-deprecations' of github.com:Cryoris/qiskit-terra into sk-deprecations

Cryoris commented:
@alexanderivrii done! And no it is still possible to generate the old `npy` format, it was pending deprecation before and is fully deprecated now.

Cryoris committed:
add concrete store + load example

this is using .. code-block:: python to avoid actually running the snippet and dumping to a local file

alexanderivrii reviewed (approved):
Thanks for adding the example on how to store/load approximations.
